File: kvx/Domain/UseCases/BuzzerUseCases.swift
import Foundation

struct BuzzerUseCases {
    let repository: any BuzzerRepository

    func load(deviceID: String) async throws -> (BuzzerDetail, [BuzzerSource], [BuzzerMotionEvent]) {
        async let detail = repository.detail(deviceID: deviceID)
        async let sources = repository.linkedPIRs(deviceID: deviceID)
        async let events = repository.history(deviceID: deviceID)
        return try await (detail, sources, events)
    }

    func availablePIRs(deviceID: String) async throws -> [AvailableBuzzerPIR] { try await repository.availablePIRs(deviceID: deviceID) }
    func link(deviceID: String, configuration: BuzzerLinkConfiguration) async throws { try await repository.link(deviceID: deviceID, configuration: configuration) }
    func unlink(deviceID: String, pirID: String) async throws { try await repository.unlink(deviceID: deviceID, pirID: pirID) }

    func test(deviceID: String) async throws -> BuzzerTestReceipt {
        try await repository.test(deviceID: deviceID)
    }
}

File: kvx_flutter/lib/application/usecases/buzzer_usecases.dart
import '../../domain/entities/buzzer_detail.dart';
import '../../domain/repositories/buzzer_repository.dart';

class BuzzerUseCases {
  final BuzzerRepository repository;
  const BuzzerUseCases(this.repository);

  Future<(BuzzerDetail, List<BuzzerSource>, List<BuzzerMotionEvent>)> load(String deviceId) async {
    final results = await Future.wait<Object>([
      repository.load(deviceId),
      repository.loadLinkedPirs(deviceId),
      repository.loadHistory(deviceId),
    ]);
    final detail = results[0] as BuzzerDetail;
    if (detail.id != deviceId) throw const BuzzerFailure(BuzzerFailureKind.invalidData);
    return (detail, results[1] as List<BuzzerSource>, results[2] as List<BuzzerMotionEvent>);
  }

  Future<BuzzerTestReceipt> test(String deviceId) => repository.test(deviceId);
}
